Astrabit-ST/Luminol
# Unsaved changes modal

**Description**
This pull request fixes several issues with project saving and adds a modal that asks the user if they want to save their current project when closing a project with unsaved changes.

The project saving has been optimized to only save files that have been modified. This is important for web builds with their slow filesystem write speed and also for projects using version control. I've added a `modified` field to each field of the data cache that indicates if the corresponding file was modified. I've also added a different `modified` field to `UpdateState` that indicates if any file in the current project was modified that the unsaved changes modal uses instead of scanning through all the individual `modified` fields in the data cache.

Currently, only the map editor and the script editor are capable of registering modifications because the item editor doesn't write changes to the data cache and the event editor is not implemented at all.

**Testing**
* Save a project, close and reopen the application, then load the project again and open any map. The application should not crash.
* Close a project using the Close Project item in the File menu in the top bar. The unsaved changes modal should appear if and only if the project was modified (i.e. a map or script was modified without saving the project afterwards).
* Open a project while a different project is loaded, using the Open Project button in the Get Started tab or the Open Project item in the File menu. The unsaved changes modal should appear if and only if the project was modified. Additionally, the original project should close before the new project is loaded.
* Open a recent project in the Get Started tab while a different project is loaded. The unsaved changes modal should appear if and only if the project was modified. Additionally, the original project should close before the new project is loaded.
* Create a new project while a different project is loaded. The unsaved changes modal should appear if and only if the project was modified.
* Close the window in a native build while a project is loaded. The unsaved changes modal should appear if and only if the project was modified.
* Close or refresh the browser tab in a web build while a project is loaded. The browser should show its builtin unsaved changes modal if and only if the project was modified.



**Checklist**

- [x] Run `cargo fmt`.
- [x] Run `cargo clippy`. If applicable, add:
  - [x] `--target wasm32-unknown-unknown`
- [x] Run `cargo build --release` 
- [x] If applicable, run `trunk build --release`
